Default HSTS storage directory is not set

pvollan opened the pull request:
#### 96a8ca8ed98d7e3f55a6c47e2f46f9aaf1f5dcac
<pre>
Default HSTS storage directory is not set
<a href="https://bugs.webkit.org/show_bug.cgi?id=243770">https://bugs.webkit.org/show_bug.cgi?id=243770</a>
&lt;rdar://problem/98432266&gt;

Reviewed by Sihui Liu.

We should provide a default HSTS storage directory in case the client is not. Reading and
writing of HSTS data will fail due to sandbox restrictions when the HSTS directory is not
set. When a default HSTS is being provided, a sandbox extension will be sent to the
WebContent process, fixing this issue. When the HSTS directory is not set, CFNetwork will
fall back to use a default path for HSTS storage. The sandbox violation is caused by the
Network process not having access to the default path.

* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
(webkit_website_data_manager_get_hsts_cache_directory):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultHSTSStorageDirectory):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::WebsiteDataStoreConfiguration):
* Source/WebKit/UIProcess/glib/WebsiteDataStoreGLib.cpp:
(WebKit::WebsiteDataStore::defaultHSTSStorageDirectory):
(WebKit::WebsiteDataStore::defaultHSTSDirectory): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/253331@main">https://commits.webkit.org/253331@main</a>
</pre>

szewai commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -298,6 +298,11 @@ static bool experimentalFeatureEnabled(const String& key, bool defaultValue = fa
     return cacheDirectoryFileSystemRepresentation("AlternativeServices"_s, ShouldCreateDirectory::No);
 }
 
+String WebsiteDataStore::defaultHSTSDirectory()
```
I think we should use `defaultHSTSStorageDirectory` to be aligned with all the other HSTSStorageDirectory names.

pvollan replied:
Fixed.

szewai commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
```diff
@@ -348,9 +348,7 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
     static WTF::String defaultApplicationCacheDirectory();
     static WTF::String defaultWebSQLDatabaseDirectory();
-#if USE(GLIB)
     static WTF::String defaultHSTSDirectory();
```
Do we need to wrap this with `#if USE(GLIB) || PLATFORM(COCOA)`?

pvollan replied:
Yes.

mcatanzaro replied:
I would remove the guards, and add a dummy implementation in WebsiteDataStore.cpp instead. Of course either way is fine.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
```diff
@@ -448,4 +448,10 @@ HTTPServer server([connectionCount = 0] (Connection connection) mutable {
+TEST(WebKit, DefaultHSTSDirectory)
+{
+    auto configuration = adoptNS([[_WKWebsiteDataStoreConfiguration alloc] init]);
+    EXPECT_GT(configuration.get().hstsStorageDirectory.absoluteString.length, 0U);
```
Can we assert the url non-null (e.g. using `EXPECT_FALSE`)?

pvollan replied:
Done!

szewai reviewed (approved):
LGTM with some nits.
I think we want to mention in the commit message that when the directory is not set, CFNetwork will fall back to use a default path for HSTS storage. Sandbox violation is caused by network process not having access to the default path.

pvollan committed:
Default HSTS storage directory is not set
https://bugs.webkit.org/show_bug.cgi?id=243770
<rdar://problem/98432266>

Reviewed by Sihui Liu.

We should provide a default HSTS storage directory in case the client is not. Reading and
writing of HSTS data will fail due to sandbox restrictions when the HSTS directory is not
set. When a default HSTS is being provided, a sandbox extension will be sent to the
WebContent process, fixing this issue. When the HSTS directory is not set, CFNetwork will
fall back to use a default path for HSTS storage. The sandbox violation is caused by the
Network process not having access to the default path.

* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
(webkit_website_data_manager_get_hsts_cache_directory):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::defaultHSTSStorageDirectory):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::WebsiteDataStoreConfiguration):
* Source/WebKit/UIProcess/glib/WebsiteDataStoreGLib.cpp:
(WebKit::WebsiteDataStore::defaultHSTSStorageDirectory):
(WebKit::WebsiteDataStore::defaultHSTSDirectory): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/253331@main
